stream_alert/rule_promotion/publisher.py

```diff
@@ -0,0 +1,143 @@
+                that are being reported on
+
+        Returns:
+            str: Entire message digest to be sent to SNS, sorted with statistics
+                that have the highest alert count at the top
+        """
+        if not stats:
+            return 'No currently staged rules to report on'
+
+        return '\n\n'.join([str(stat) for stat in sorted(stats, reverse=True)])
+
+    @classmethod
+    def _load_state(cls):
+        """Retrieve the remote state configuration from systems manager
+
+        Returns:
+            dict: The JSON configuration loaded into a dictionary
+        """
+        cls.SSM_CLIENT = boto3.client('ssm')
+        response = cls.SSM_CLIENT.get_parameter(Name=cls.SSM_STATE_NAME)
+        return json.loads(response['Parameter']['Value'])
+
+    def _write_state(self):
+        """Write the new state configuration to systems manager"""
+        self._state['sent_daily_digest'] = (
+            self._state['send_digest_hour_utc'] == self._current_time.hour
+        )
+        param_value = json.dumps(self._state)
+        StatsPublisher.SSM_CLIENT.put_parameter(
+            Name=self.SSM_STATE_NAME,
+            Value=param_value,
+            Overwrite=True
+        )
+
+    def _query_alerts(self, stat):
+        """Execute a query for all alerts for a rule so the user can be sent the results
+
+        Args:
+            rule_name (str): Name of the rule to query for alert results
+
+        Returns:
+            str: Execution ID for running Athena query
+        """
+        # If there are no alerts, do not run the comprehensive query
+        if not stat.alert_count:
+            return
+
+        info_statement = stat.sql_info_statement
+        LOGGER.debug('Querying alert info for rule \'%s\': %s', stat.rule_name, info_statement)
+
+        response = self._athena_client.run_async_query(info_statement)
+        if not response:
+            LOGGER.error('Failed to query alert info for rule: \'%s\'', stat.rule_name)
+            return
+
+        return response['QueryExecutionId']
+
+    def _publish_message(self, stats):
+        """Publish the alert statistics message to SNS
+
+        Args:
+            stats (list<StagingStatistic>): Group of rule staging statistics
+                that are being reported on
+        """
+        LOGGER.info('Sending daily message digest at %s', self._current_time)
+
+        sns_client = boto3.resource('sns').Topic(self._topic_arn)
+
+        subject = 'Alert statistics for {} staged rule(s) [{} UTC]'.format(
+            len(stats),
+            self._current_time
+        )
+
+        sns_client.publish(
+            Message=self._format_digest(stats),
+            Subject=subject
+        )
+
+    @property
+    def _should_send_digest(self):
+        """Returns True if the daily digest should be sent, False otherwise"""
+        return (
+            self._state['send_digest_hour_utc'] == self._current_time.hour
```

Comment:
It doesn't look like `self._state['send_digest_hour_utc']` is ever updated?  It's not modified in `_write_state`

It might be simpler to simply mark the last time a digest was sent

Reply:
the 'send_digest_hour_utc' is read-only and shouldn't be changed/written from the function. This is the 'hour of the day at which the daily digest should be sent'.. it's not obviously since this is a work in progress, but this value will come from the user's config and be written to systems manager.

How would marking the last time it was sent help with scheduling the next one to be sent? Remember, this function will run often (every 10 minutes) and not only once a day. I'm open to suggestions as to how to keep it from sending more than one digest a day if you've got creative ideas :)

Reply:
@austinbyers setting a last sent timestamp is not a great solution, since this will run on an interval - that means if it runs every 15 minutes and just checks to see if this execution is _after_ the previous sent time - the digest will be sent later and later each day. this is my main concern. it needs to send at _roughly_ the same time each day or as close to it as possible, and not get progressively later

Reply:
Oh ok got it, I didn't realize this was going to be a hardcoded time-of-day thing 👍

Reply:
just the digest sending part. I'm not sure if will stay this way but we'll go with it for now :)